kernel/src/table_changes/schema_compat.rs

```diff
@@ -0,0 +1,302 @@
+use std::collections::{HashMap, HashSet};
+
+use crate::schema::{DataType, Schema, StructField, StructType};
+
+fn is_nullability_compatible(existing_nullable: bool, read_nullable: bool) -> bool {
+    // The case to avoid is when the read_schema is non-nullable and the existing one is nullable.
+    // So we avoid the case where !read_nullable && existing_nullable
+    // Hence we check that !(!read_nullable && existing_nullable)
+    // == read_nullable || !existing_nullable
+    read_nullable || !existing_nullable
+}
+fn is_struct_read_compatible(existing: &StructType, read_type: &StructType) -> bool {
+    // Delta tables do not allow fields that differ in name only by case
+    let existing_fields: HashMap<&String, &StructField> = existing
+        .fields()
+        .map(|field| (field.name(), field))
+        .collect();
+
+    let existing_names: HashSet<String> = existing
+        .fields()
+        .map(|field| field.name().clone())
+        .collect();
+    let read_names: HashSet<String> = read_type
+        .fields()
+        .map(|field| field.name().clone())
+        .collect();
+    if !existing_names.is_subset(&read_names) {
+        return false;
+    }
+    read_type
+        .fields()
+        .all(|read_field| match existing_fields.get(read_field.name()) {
+            Some(existing_field) => {
+                let name_equal = existing_field.name() == read_field.name();
+
+                let nullability_equal =
+                    is_nullability_compatible(existing_field.nullable, read_field.nullable);
+                let data_type_equal =
+                    is_datatype_read_compatible(existing_field.data_type(), read_field.data_type());
+                println!(
+                    "name_equal {} nullability: {}, datatype: {}",
+                    name_equal, nullability_equal, data_type_equal
+                );
+                name_equal && nullability_equal && data_type_equal
+            }
+            None => read_field.is_nullable(),
```

Comment:
The `None` case is a point at which I differ from the delta implementation. I'm [not convinced by the code there](https://github.com/delta-io/delta/blob/4d2c5cfd45119d78a86560de12e7b946228d77f2/spark/src/main/scala/org/apache/spark/sql/delta/schema/SchemaUtils.scala#L454). If we don't find the read field in the existing schema, then we just ignore it. I think this should only pass if the new field in the read schema is nullable.

I may be missing something tho 🤔